src/plugins/discord/sqliteMirror.js

```diff
@@ -497,4 +497,17 @@ export class SqliteMirror {
       .all({channel_id: channel, message_id: message})
       .map((e) => Model.refToEmoji(e.emoji));
   }
+
+  latestMessageId(channel: Model.Snowflake) {
+    return this._db
+      .prepare(
+        dedent`
+          SELECT id, max(timestamp_ms)
+          FROM messages
+          WHERE channel_id = :channel_id
```

Comment:
Sorting by `timestamp_ms` is technically imperfect. Because the real order is determined by sorting Snowflakes treated as `uint64`. The only time these are different is when the timestamp is identical for several entries, and the Snowflakes' additional data need to act as a tiebreaker.
https://discord.com/developers/docs/reference#snowflakes

More importantly, in the prototype what I ended up using was not a query for the last message but:
https://github.com/sourcecred/sourcecred/blob/d0dc11d2016dcb359606ee736a89e20235a7aaaf/src/plugins/discord/mirrorRepository.js#L208

The reason being, we'll want to "reload" some amount of messages we already cached. Because they could be edited (for example adding new mentions, changing the graph) and the emoji reactions on them may have changed since last mirror. The number of messages is a heuristic, trading off between faster mirroring and higher accuracy.

We don't care too much that the sorting is imperfect there, because the number `n` was arbitrary anyway. If we reload one message more or less, it wouldn't be a big issue. So I went with `timestamp_ms` sorting there.

---

Concluding, it would be nice-to-have correct sorting. But more importantly, want to check if `latestMessageId` is what's needed for our caching strategy or whether we'll be using `nthMessageFromTail` like queries instead?

Reply:
It feels like `nthMessageFromTail` will be more useful; I'm wondering if `n` is the appropriate parameter or if we should use something like time (say, all messages in the last five days) to re-fetch? 

E.g., 50 messages can take a month or a few minutes to accumulate in Discord. However, once messages are buried behind others, it's much less likely they'll be updated. 

What do you think?

Reply:
Good idea, sounds like that would work better than Nth.

Say `n=30` I can definitely imagine a lot of unnecessary reloading of something like a locked down `#announcements` channel where messages trickle in every other week. It could mean reloading the whole channel. Versus possibly not enough reloading for high volumes of messages in something like `#support` or `#general` where 30 messages could be 5 minutes worth of conversation.

This burying of messages you're referring to is also the underlying assumption why Nth message would still gave acceptable accuracy. I think that's also true for a time offset :]

Depending on what acceptable accuracy means, you could tweak the parameter or discard the cache whenever you like. For example, using cache for updates of stats several times a day but reloading everything just before the weekly Grain distribution.

Reply:
Happy to accept either approach :]
Whether it holds up is we'll probably only find out by testing it in the wild.